ShowEmptyFields.js

```diff
@@ -0,0 +1,372 @@
+                    || this.emptyTabCopy[i] === arrayLeftFinal[j] + board.x) {
+                  // same logic as before but without -1(so it wouldn't open right fields)
+                  arrayLeftFinal.push(this.emptyTabCopy[i]);
+                }
+              }
+            }
+          }
+        }
+        arrayLeftFinal = arrayLeftFinal.filter((item, index) => arrayLeftFinal.indexOf(item) === index);
+        outArray.push(arrayLeftFinal);
+        arrayLeftFinal = [];
+      }
+    };
+    arraysForDifferentSides(this.tempArrayLeftSeparate, this.leftArraysDone, 0);
+    arraysForDifferentSides(this.tempArrayRightSeparate, this.rightArraysDone, 1);
+    doublesDelete(this.leftArraysDone);
+    doublesDelete(this.rightArraysDone);
+    this.array1 = this.array1.filter((item, index) => this.array1.indexOf(item) === index);
+    // deleting doubles for the next array
+    this.emptyTabCopy = this.emptyTabCopy.filter((val) => this.array1.indexOf(val) === -1);
+    // adding our left and right arrays to the final array
+    for (let i = 0; i < this.leftArraysDone.length; i++) {
+      this.tempArray.push(this.leftArraysDone[i]);
+    }
+    for (let i = 0; i < this.rightArraysDone.length; i++) {
+      this.tempArray.push(this.rightArraysDone[i]);
+    }
+    this.tempArray.push(this.array1);
+    doublesDelete(this.tempArray);
+    this.array1 = [];
+    if (this.emptyTabCopy.length !== 0) {
+      this.showFieldsWithNoMinesAround(); // using recursion to create arrays with empty fields
+      // around mines until array with all empty fields is empty
+    }
+  }
+
+  searchForDuplicatesInArray() {
+    for (let i = 0; i < this.tempArray.length; i++) { // comparing array and searching
+      // for the same numbers because there could be arrays
+      // which supposed to be one but were separated
+      for (let j = 0; j < this.tempArray[i].length; j++) {
+        for (let k = 0; k < this.tempArray.length; k++) {
+          this.tempArray[i] = this.tempArray[i].filter((item, index) => this.tempArray[i].indexOf(item) === index);
+          for (let l = 0; l < this.tempArray[k].length; l++) {
+            if (i !== k) {
+              // if any same number is found we are combining two arrays and delete one of them
+              if (this.tempArray[i][j] === this.tempArray[k][l]) {
+                this.tempArray[i] = this.tempArray[i].concat(this.tempArray[k]);
+                this.tempArray.splice(this.tempArray.indexOf(this.tempArray[k]), 1);
+                k--; // undefined k length fix
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  convertEmptyFieldsArrayToString() {
+    // converting numbers in the array back to string
+    for (let i = 0; i < this.tempArray.length; i++) {
+      let tempVar = this.tempArray[i];
+      tempVar = tempVar.map(String);
+      this.emptyTabCopy.push(tempVar);
+    }
+  }
+
+  fixEmptyFields() {
```

Comment:
this function as well is to complicated. You need to write simple code. Having 4 for loops one inside the other that's to much.